infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/LogicalGroupSnapshotReader.java

```diff
@@ -1,21 +1,180 @@
+                    if (currentStreamInfo.iterator.hasNext()) {
+                        break;
+                    } else {
+                        // Skip process this stream as it has no entries to process, will poll the next one.
+                        log.info("Snapshot reader will skip reading stream {} as there are no entries to send",
+                                currentStreamInfo.uuid);
+                    }
+                } else {
+                    log.info("Sending CLEAR_ENTRY for stream name={}, id={}", streamName,
+                            CorfuRuntime.getStreamID(streamName));
+                    isReadingStreamToClear = true;
+                    messages.add(generateClearMessage(streamName, syncRequestId));
+                    break;
+                }
+            }
+        }
+
+        if (currentStreamHasNext()) {
+            msg = read(currentStreamInfo, syncRequestId);
+            if (msg != null) {
+                messages.add(msg);
+            }
+        }
+
+        if (!currentStreamHasNext()) {
+            log.debug("Snapshot log reader finished reading stream id={}, name={}", currentStreamInfo.uuid,
+                    currentStreamInfo.name);
+            currentStreamInfo = null;
+
+            if (streamsToSend.isEmpty()) {
+                log.info("Snapshot log reader finished reading ALL streams, total={}", streams.size());
+                endSnapshotSync = true;
+            }
+        }
+
+        isReadingStreamToClear = false;
+        return new SnapshotReadMessage(messages, endSnapshotSync);
+    }
+
+    LogReplication.LogReplicationEntryMsg generateClearMessage(String streamName, UUID snapshotRequestId) {
+        //mark the end of the current stream as there is only 1 CLEAR_ENTRY to send.
+        currentMsgTs = snapshotTimestamp;
+        LogReplication.LogReplicationEntryMetadataMsg metadata = LogReplication.LogReplicationEntryMetadataMsg
+                .newBuilder()
+                .setEntryType(LogReplication.LogReplicationEntryType.SNAPSHOT_MESSAGE)
+                .setTopologyConfigID(topologyConfigId)
+                .setSyncRequestId(getUuidMsg(snapshotRequestId))
+                .setTimestamp(currentMsgTs)
+                .setPreviousTimestamp(preMsgTs)
+                .setSnapshotTimestamp(snapshotTimestamp)
+                .setSnapshotSyncSeqNum(sequence)
+                .build();
+
+        Map<UUID, List<SMREntry>> map = new HashMap<>();
+        map.put(CorfuRuntime.getStreamID(streamName), Collections.singletonList(CLEAR_ENTRY));
+        OpaqueEntry opaqueEntry = new OpaqueEntry(0, map);
+
+        LogReplication.LogReplicationEntryMsg txMsg = getLrEntryMsg(unsafeWrap(generatePayload(opaqueEntry)), metadata);
+
+        preMsgTs = currentMsgTs;
+        sequence++;
+
+        return txMsg;
+    }
+
+    @Override
+    boolean currentStreamHasNext() {
+        if (isReadingStreamToClear) {
```

Comment:
instead of changing currentStreamHasNext(), it will be cleaner to check for 'isReadingStreamToClear' in the calling read() method only

Reply:
modified, thanks!